src/media/class-wpml-beaver-builder-media-node-provider.php

```diff
@@ -0,0 +1,50 @@
+<?php
+
+class WPML_Beaver_Builder_Media_Node_Provider {
+
+	/** @var WPML_Page_Builders_Media_Translate_Helper $translate_helper */
+	private $translate_helper;
+
+	/** @var WPML_Beaver_Builder_Media_Node[] */
+	private $nodes = array();
+
+	/**
+	 * @param string $type
+	 *
+	 * @return WPML_Beaver_Builder_Media_Node
+	 */
+	public function get( $type ) {
+		if ( ! array_key_exists( $type, $this->nodes ) ) {
+			switch ( $type ) {
+				case 'photo':
+					$node = new WPML_Beaver_Builder_Media_Node_Photo( $this->get_translation_helper() );
+					break;
+
+				case 'gallery':
+					$node = new WPML_Beaver_Builder_Media_Node_Gallery( $this->get_translation_helper() );
+					break;
+
+				default:
+					$node = null;
+			}
+
+			$this->nodes[ $type ] = $node;
+		}
+
+		return $this->nodes[ $type ];
+	}
+
+	/** @return WPML_Page_Builders_Media_Translate_Helper */
+	private function get_translation_helper() {
+		global $sitepress;
+
+		if ( ! $this->translate_helper ) {
+			$this->translate_helper = new WPML_Page_Builders_Media_Translate_Helper(
+				new WPML_Translation_Element_Factory( $sitepress ),
+				new WPML_Media_Image_Translate( $sitepress, new WPML_Media_Attachment_By_URL_Factory() )
+			);
+		}
+
+		return $this->translate_helper;
+	}
+}
```

Comment:
@strategio Technically all this is correct but is all this storing of class instances needed. This whole class could be replaced with the following and I expect it would have zero impact on performance.

```
<?php
class WPML_Beaver_Builder_Media_Node_Provider {
	
	/**
	 * @param string $type
	 *
	 * @return WPML_Beaver_Builder_Media_Node
	 */
	public function get( $type ) {
		switch ( $type ) {
			case 'photo':
				$node = new WPML_Beaver_Builder_Media_Node_Photo( $this->get_translation_helper() );
				break;
			case 'gallery':
				$node = new WPML_Beaver_Builder_Media_Node_Gallery( $this->get_translation_helper() );
				break;
			default:
				$node = null;
		}
		return $node;
	}
	
	/** @return WPML_Page_Builders_Media_Translate_Helper */
	private function get_translation_helper() {
		global $sitepress;
		return new WPML_Page_Builders_Media_Translate_Helper(
				new WPML_Translation_Element_Factory( $sitepress ),
				new WPML_Media_Image_Translate( $sitepress, new WPML_Media_Attachment_By_URL_Factory() )
			);
	}
}
```

Reply:
@brucepearson, the translation helper will cache translation elements so I thought it could reduce overhead if we reuse the same instance (see in https://github.com/OnTheGoSystems/wpml-page-builders/blob/wpmlpb-160-part2/src/media/class-wpml-page-builders-media-translate-helper.php).